### PR TITLE
Parse bare function type

### LINKED ISSUE #961: Parse bare `function` type
We need to recognize `function` by itself as meaning any function.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Parse bare function type
Resolves #961 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: runtime/balrt.h [hunk 1]
```diff
@@ -346,6 +346,11 @@ typedef GC struct LargeString {
     char bytes[];
 } *LargeStringPtr;
 
+typedef void *FunctionPtr;
```
Comment (jclark) [inline runtime/balrt.h]: I would make this a function pointer

Comment (heshanpadmasiri) [reply runtime/balrt.h]: Fixed

File: runtime/balrt.h [hunk 2]
```diff
@@ -346,6 +346,11 @@ typedef GC struct LargeString {
     char bytes[];
 } *LargeStringPtr;
 
+typedef void(*FunctionPtr)();
```
Comment (jclark) [inline runtime/balrt.h]: Sace after void

File: docs/subset15.md [hunk 3]
```diff
@@ -403,9 +404,14 @@ Two kinds of `import` are supported.
 
 * `list-constructor-expr` and `mapping-constructor-expr` are not allowed within a `const-expr`.
 * Types in type definitions are restricted semantically, rather than syntactically: a type definition that is referenced from a function definition must define a type that is equivalent to one that can be described using the type-defn grammar in this document. It must also match the type-defn [grammar supported for semantic type-checking](type-subset.md).
+* Structure types can not have subtypes of `function` as a member type.
+* The type `T` in type test expression `E is T` or `E !is T` can not be a proper subtype of `function`.
+* The type `T` in type case expression `<T>E` can not be a proper subtype of `function`.
```
Comment (jclark) [inline docs/subset15.md]: cast

File: docs/subset15.md [hunk 4]
```diff
@@ -403,9 +404,14 @@ Two kinds of `import` are supported.
 
 * `list-constructor-expr` and `mapping-constructor-expr` are not allowed within a `const-expr`.
 * Types in type definitions are restricted semantically, rather than syntactically: a type definition that is referenced from a function definition must define a type that is equivalent to one that can be described using the type-defn grammar in this document. It must also match the type-defn [grammar supported for semantic type-checking](type-subset.md).
+* Structure types can not have subtypes of `function` as a member type.
```
Comment (jclark) [inline docs/subset15.md]: Put all these under a single bullet "Function typing is not yet fully implemented"

File: docs/subset15.md [hunk 5]
```diff
@@ -20,6 +20,7 @@ For programs that are in the subset, the compiler should conform to the Ballerin
    * `any` type
    * `byte` type
    * `anydata` type
+   * `function` type
```
Comment (jclark) [inline docs/subset15.md]: Bullet below with "function types" with two sub bullets (one for bare function type)

File: runtime/balrt.h (hunk 2 above)
Comment (heshanpadmasiri) [reply runtime/balrt.h]: fixed

File: docs/subset15.md (hunk 3 above)
Comment (heshanpadmasiri) [reply docs/subset15.md]: fixed

File: docs/subset15.md (hunk 4 above)
Comment (heshanpadmasiri) [reply docs/subset15.md]: Fixed

File: docs/subset15.md (hunk 5 above)
Comment (heshanpadmasiri) [reply docs/subset15.md]: Fixed

**Review by jclark — APPROVED** [review]

**Pull request merged**